### PR TITLE
Enable nullable on AttachVS

### PR DESCRIPTION
**Pull request by Evangelink**
Contributes to [AB#1549371](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1549371)

### CODE REVIEW

File: src/AttachVS/AttachVs.cs [hunk 1]
```diff
@@ -215,18 +218,18 @@ private static bool IsVsOrNull(Process process)
         return isVs;
     }
 
-    private static bool IsCorrectParent(Process currentProcess, Process parent)
+    private static bool IsCorrectParent(Process currentProcess, Process? parent)
```
Comment (Evangelink) [inline src/AttachVS/AttachVs.cs:221]: We could also ensure that `parent` is not nullable and handle potential nulls on calling sites. Any preference?

File: src/AttachVS/AttachVs.cs [hunk 2]
```diff
@@ -38,20 +36,25 @@ internal static bool AttachVSToProcess(int? pid, int? vsPid)
...
+                            return null;
+                        }
+                    })
+                    .Where(p => p != null && !p.HasExited)
+                    .OrderBy(p => p!.StartTime)
+                    .ToList();
 
-                var firstVs = processes.FirstOrDefault();
-                Trace($"Found VS {firstVs.Process.Id}");
+                var firstVs = processes.First();
```
Comment (Evangelink) [inline src/AttachVS/AttachVs.cs:56]: We were using `FirstOrDefault` but then doing member access. So it's unclear to me whether or not we expect to have an empty list.

Comment (nohwnd) [reply src/AttachVS/AttachVs.cs:56]: I will have to look at the logic again. Plan was originally to: pickup the "current" process from env variable, or use the current. Then for that  pickup parent VS, and if not found fallback to first VS started. And if not found then fail.

File: src/AttachVS/AttachVs.cs [hunk 3]
```diff
@@ -215,61 +223,51 @@ private static bool IsVsOrNull(Process process)
...
+                return p;
             }
             catch
             {
-                // throws when parent no longer runs
+                return -1;
             }
         }
-
-        return IsCorrectParent(process, parent) ? parent : null;
```
Comment (nohwnd) [inline src/AttachVS/AttachVs.cs:269]: Why did this logic have to move into the body of the other method directly? Separating it into a named method seemed clearer.

File: src/AttachVS/AttachVs.cs [hunk 4]
```diff
@@ -215,61 +223,51 @@ private static bool IsVsOrNull(Process process)
...
-            id = p;
-        }
-        catch
-        {
-            id = -1;
+                Trace($"Process {parent.ProcessName} ({parent.Id}) is not a valid parent because it started after the current process.");
+            }
+            catch
+            {
+                // throws when parent no longer runs
```
Comment (nohwnd) [inline src/AttachVS/AttachVs.cs:246]: This catch is now not as narrow scoped as before, it can throw for more reasons I think.

Comment (Evangelink) [reply src/AttachVS/AttachVs.cs:246]: All the `IsCorrect` was inside a global try/catch and the call to `GetProcessById` was also inside a global try/catch so I have simply merged the 2.

File: src/AttachVS/AttachVs.cs (hunk 3 above)
Comment (Evangelink) [reply src/AttachVS/AttachVs.cs:269]: Seems more complex to have all the logic split but I will revert.

Separated looks like:

image

Merged:

image

**Review by nohwnd — APPROVED** [review]

**Pull request merged**